src/clang-c-frontend/clang_c_convert.cpp

```diff
@@ -333,21 +330,38 @@ bool clang_c_convertert::get_struct_union_class(const clang::RecordDecl &rd)
   // We have to add the struct/union/class to the context before converting its
   // fields because there might be recursive struct/union/class (pointers) and
   // the code at get_type, case clang::Type::Record, needs to find the correct
-  // type (itself). Note that the type is incomplete at this stage, it doesn't
-  // contain the fields, which are added to the symbol later on this method.
-  move_symbol_to_context(symbol);
+  // type (itself). That's why the symbol's type is symbolic for this step.
+  /* The downside of this approach is that we can't use ns.follow() until this
+   * symbol (and all type symbols up our callstack) are properly constructed:
+   * it would loop forever. */
+
+  sym = move_symbol_to_context(symbol);
+
+  // TODO: Fix me when we have a test case using C++ union.
+  //       A C++ union can have member functions but not virtual functions.
+  //       Just use struct_typet for C++?
+  struct_union_typet t(rd.isUnion() ? "union" : "struct");
+  t.tag(name);
+  t.location() = location_begin;
 
   // Don't continue to parse if it doesn't have a complete definition
   // Try to get the definition
   clang::RecordDecl *rd_def = rd.getDefinition();
   if(!rd_def)
+  {
+    t.id("incomplete_" + t.id().as_string());
```

Comment:
do we need to change the id here? I'd assume setting incomplete(true) is enough

Reply:
Unfortunately not. `type.is_struct()` or `type.id() == "struct"` (same for union) is used often and many of those won't work for incomplete structures. Before this PR, `incomplete()` was completely unused, even the old frontend didn't use it. So no, just setting incomplete(true) won't work. The rest of the code is just not expecting it.

Reply:
by rest of the code, you mean only adjust, right? Since there should be no incomplete type when we start goto conversion, either they are complete or they are symbolic_typet.

Reply:
I don't mean adjust only. Everything else. Until SMT. Look at migrate_type() for instance, there is a case about incomplete_struct. It might be a remnant from the good old days™, but still: in my opinion an incomplete type (one that just has been declared, but never defined) is different from a symbolic type (which is basically a typedef). A pointer to an incomplete type can never be dereferenced. On the other hand a pointer to a symbolic type *could be*, depending on whether ns.follow(sym_type) is complete or not.

Reply:
I have been too quick to reply, sorry. It's not until SMT, just until c_link. And we'll probably need it until there in order for C++ types to be resolved across translation units. Nonetheless, I still believe incomplete type != symbolic type.